[DDCI-124] - added vocab auto complete

awset opened the pull request:
https://it-partners.atlassian.net/browse/DDCI-124

@salsa-nathan there are issues here.
how it is work based on the solution direction
- text field becomes autocomplete
- filter by the label, and the value is URI as per other vocab.

issue:
- when there is an **error** or **on edit form,** the value become URI instead of the label, 
I tried to add a hidden field to hold the value and render the label, etc, but still got issue when there is an error on the new dataset form. and I haven't found a solution for this. 

and if we can change the field to dropdown, and dynamically have the dropdown updated based on the ajax, it seems there is a plenty of work, since I am not sure if the js plugin is supporting this OOTB.


![image](https://user-images.githubusercontent.com/1538818/94913777-08ac0080-04d4-11eb-99ee-22c298e87bda.png)

on edit/error form
![image](https://user-images.githubusercontent.com/1538818/94913817-1e212a80-04d4-11eb-939d-908d6e3d021b.png)

FE
![image](https://user-images.githubusercontent.com/1538818/94913840-2c6f4680-04d4-11eb-8733-fcafc33c265a.png)

example field
```  
  {
      "field_name": "test_field",
      "label": "Test Field",
      "preset": "qdes_vocab_service_autocomplete",
      "vocabulary_service_name": "spatial_name_code",
      "form_placeholder": "Select a Test Field",
      "form_attrs": {
        "data-module": "qdes_autocomplete",
        "data-module-source": "/ckan-admin/vocabulary-service/term-autocomplete/spatial_name_code?incomplete=?",
        "data-module-key": "value",
        "data-module-label": "label"
      }
    },
```

awset committed:
[DDCI-124] - added vocab auto complete

MarkCalvert commented:
Hey @awset @salsa-nathan,
I also had this issue using the autocomplete API for the Related and lineage metadata fields.
The way I got around this issue with was to store the whole selected object data as the value eg. `{"id":dataset_id, "text":dataset_title}` in JS `value = $(this).select2('data');`.
If there is an error on the form the data object above is sent back to the form, and I added some code in the 'autocomplete.js' file to create the initialValue from the object value https://github.com/salsa-nathan/ckanext-qdes-schema/blob/develop/ckanext/qdes_schema/fanstatic/autocomplete.js#L281.
This is quite a big change though as we currently only store the value (URL for vocab services) and will need to add some JS to get the selected object value just in case their is a form error.
We then have a potential problem of other code expecting just the URL value instead of an object so we need to think about if we should convert this object back into the URL value in the package interface 'after_update' or a converter.

We can discuss options on Tuesday.

salsa-nathan commented:
> Hey @awset @salsa-nathan,
> I also had this issue using the autocomplete API for the Related and lineage metadata fields.
> The way I got around this issue with was to store the whole selected object data as the value eg. `{"id":dataset_id, "text":dataset_title}` in JS `value = $(this).select2('data');`.
> If there is an error on the form the data object above is sent back to the form, and I added some code in the 'autocomplete.js' file to create the initialValue from the object value https://github.com/salsa-nathan/ckanext-qdes-schema/blob/develop/ckanext/qdes_schema/fanstatic/autocomplete.js#L281.
> This is quite a big change though as we currently only store the value (URL for vocab services) and will need to add some JS to get the selected object value just in case their is a form error.
> We then have a potential problem of other code expecting just the URL value instead of an object so we need to think about if we should convert this object back into the URL value in the package interface 'after_update' or a converter.
> 
> We can discuss options on Tuesday.

@MarkCalvert - did you get a chance to discuss this with @awset yet?

MarkCalvert commented:
> > Hey @awset @salsa-nathan,
> > I also had this issue using the autocomplete API for the Related and lineage metadata fields.
> > The way I got around this issue with was to store the whole selected object data as the value eg. `{"id":dataset_id, "text":dataset_title}` in JS `value = $(this).select2('data');`.
> > If there is an error on the form the data object above is sent back to the form, and I added some code in the 'autocomplete.js' file to create the initialValue from the object value https://github.com/salsa-nathan/ckanext-qdes-schema/blob/develop/ckanext/qdes_schema/fanstatic/autocomplete.js#L281.
> > This is quite a big change though as we currently only store the value (URL for vocab services) and will need to add some JS to get the selected object value just in case their is a form error.
> > We then have a potential problem of other code expecting just the URL value instead of an object so we need to think about if we should convert this object back into the URL value in the package interface 'after_update' or a converter.
> > We can discuss options on Tuesday.
> 
> @MarkCalvert - did you get a chance to discuss this with @awset yet?

@salsa-nathan No I haven't, let's discuss after stand up. @awset

salsa-nathan committed:
[DDCI-124] Refactoring `vocabulary_service_term_search`
- Adjusted search query to be case insensitive
- Refactored code for simplicity and readability
